raft/src/main/java/org/apache/kafka/raft/KafkaRaftClient.java

```diff
@@ -329,8 +349,15 @@ private VoteResponseData handleVoteRequest(VoteRequestData request) throws IOExc
             } else if (state.hasVoted()) {
                 voteGranted = state.isVotedCandidate(candidateId);
             } else {
-                OffsetAndEpoch lastEpochEndOffset = new OffsetAndEpoch(request.lastEpochEndOffset(), request.lastEpoch());
-                voteGranted = lastEpochEndOffset.compareTo(endOffset()) >= 0;
+                int lastEpoch = request.lastEpoch();
+                long lastEpochEndOffset = request.lastEpochEndOffset();
+                if (lastEpochEndOffset < 0 || lastEpoch < 0 || lastEpoch >= request.candidateEpoch()) {
```

Comment:
I was considering this. The checks in `handleFetchQuorumRecords` and `handleVoteRequest` are slightly different. We can be stricter in `handleVoteRequest` because a candidate cannot have any appended entries from the epoch of its candidacy.